### PR TITLE
21398: Updates the Amalgam unit test framework for ease of use

### PR DESCRIPTION
**Pull request by cademack**
Changes the unit test framework in the following ways:

- Adds strict checks for the `exp` variable for assert methods that use it. Calling these methods without setting the a non-null value for `exp` will now trigger a failure
- Updated the value printing logic to always print values from a failed assert if `#print_values_on_fail` is `(true)` which is on by default and switches the previous `#verbose_on_fail` which caused all assert values to be printed after a single failure to now be `(false)` by default.
- Parametrizes the sorting logic in `#assert_approximate`
- Some minor style/comment changes

Also updates the type schema and a unit test as "familiarity_conviction" is not actually a functional react detail.

### CODE REVIEW

File: unit_tests/unit_test.amlg [hunk 1]
```diff
@@ -214,188 +238,234 @@
...
-			(print msg)
-			(if (= 2 verbose_mode)
-				(print " at\n" (stack)) ;TODO: figure out a less verbose/more concise way to give feedback as to where exactly the failure occured
-			)
+	(seq
+		(assign_to_entities (assoc num_failed_asserts (+ 1 num_failed_asserts)))
+		(assign (assoc failed_assert (true) ))
+		(print msg)
+		(if (= 2 verbose_mode)
+			(print " at\n" (stack)) ;TODO: figure out a less verbose/more concise way to give feedback as to where exactly the failure occured
```
Comment (howsohazard) [inline unit_tests/unit_test.amlg]: TODOs need an associated ticket number

File: unit_tests/unit_test.amlg [hunk 2]
```diff
@@ -37,33 +40,39 @@
...
+		(call assert_true (assoc obs (!= (null) unit_test_name)))
+		(call assert_false (assoc obs (= (null) unit_test_name)))
+		(call exit_if_failures (assoc quiet_pass (true) msg "Unit test framework Failed to load!"))
 
-			(print "=====================================\n")
-		)
+		(print "=====================================\n")
+	)
+
+	;helper function that gives failure output when a non-null value is not given for exp, skips check if called within a recursive assertion
```
Comment (howsoRes) [inline unit_tests/unit_test.amlg]: 'gives' is a bit too generic, maybe '... that prints failure output ...' ?
also if this is a helper method that shouldn't be directly called by unit tests, make it private with a preceding !
also consider doing same for other helper/internal-only methods?

File: unit_tests/unit_test.amlg [hunk 3]
```diff
@@ -214,188 +238,234 @@
...
 
 	;asserts that the obs parameter is null
 	; obs : item to check if null
 	#assert_null
-		(seq
-			(if (= (null) obs)
-				(print "PASSED (null)")
+	(declare
+		(assoc
+			obs "You did not pass in a value for obs" ;intentionally not null
```
Comment (howsoRes) [inline unit_tests/unit_test.amlg]: comment on same line?

Comment (cademack) [reply unit_tests/unit_test.amlg]: ah right

File: unit_tests/unit_test.amlg (hunk 1 above)
Comment (cademack) [reply unit_tests/unit_test.amlg]: just gonna remove the comment honestly, this doesn't seem necessary

**Review by howsoRes — APPROVED** [review]

**Pull request merged**